File: src/main/java/de/bonndan/nivio/output/map/TilePath.java
package de.bonndan.nivio.output.map;

import de.bonndan.nivio.output.map.hex.Hex;

import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;

class TilePath {

    List<Hex> tiles = new ArrayList<>();
    private final List<Hex> bends = new ArrayList<>();
    private boolean closed = false;

    TilePath(Hex tile) {
        if (tile != null)
            this.tiles.add(tile);
    }

    void close() {
        if (this.closed)
            return;

        this.closed = true;

        var i = 0;
        for (i = 1; i < this.tiles.size() - 1; i++) {
            var prev = this.tiles.get(i - 1);
            var cur = this.tiles.get(i);
            var next = this.tiles.get(i + 1);
            var qBend = (prev.q == cur.q && next.q != cur.q) || (prev.q != cur.q && next.q == cur.q);
            var rBend = (prev.r == cur.r && next.r != cur.r) || (prev.r != cur.r && next.r == cur.r);
            if (qBend || rBend)
                this.bends.add(cur);
        }

    }

    void reducePoints() {
        //reduce tiles (not first or last)


        /*
        var reduced = [];
        console.log(drop);
        for (i = 0; i < this.tiles.length; i++) {
            if (drop.indexOf(i) !== -1)
                continue;
            reduced.push(this.tiles[i]);
        }
        this.tiles = reduced;
         */
    }

    String getPoints() {
        String points = "M";
        for (var i = 0; i < tiles.size(); i++) {
            var hex = tiles.get(i);
            if (this.isBend(hex)) {
                //cubic curve
                var prev = tiles.get(i - 1).toPixel();
                var point = hex.toPixel();
                var next = tiles.get(i + 1).toPixel();

                var newBefore = new Point2D.Double();
                newBefore.x = prev.x + (point.x - prev.x) / 2;
                newBefore.y = prev.y + (point.y - prev.y) / 2;
                points += " " + newBefore.x + "," + newBefore.y + " ";
                points += "Q " + point.x + "," + point.y + " ";

                var newAfter = new Point2D.Double();
                newAfter.x = next.x + (point.x - next.x) / 2;
                newAfter.y = next.y + (point.y - next.y) / 2;
                points += " " + newAfter.x + "," + newAfter.y + " L";
            } else {
                var p = hex.toPixel();
                points += " " + p.x + "," + p.y + " L";
            }
        }
        return points.substring(0, points.length() - 1);
    }

    private boolean isBend(Hex hex) {

        for (Hex bend : this.bends) {
            if (bend.q == hex.q && bend.r == hex.r) {
                return true;
            }
        }

        return false;
    }

    int getSpeed() {
        return this.tiles.size() - this.bends.size();
    }

    boolean isClosed() {
        return closed;
    }
}

File: src/main/app/src/hexmap/TilePath.js
import { HexUtils } from 'react-hexgrid';

class TilePath {
  constructor(tile) {
    this.tiles = [];
    this.bends = [];
    if (tile !== undefined) this.tiles.push(tile);
    this.closed = false;
  }

  close() {
    if (this.closed) return;

    this.closed = true;

    let i = 0;
    for (i = 1; i < this.tiles.length - 1; i++) {
      let prev = this.tiles[i - 1];
      let cur = this.tiles[i];
      let next = this.tiles[i + 1];
      let qBend = (prev.q === cur.q && next.q !== cur.q) || (prev.q !== cur.q && next.q === cur.q);
      let rBend = (prev.r === cur.r && next.r !== cur.r) || (prev.r !== cur.r && next.r === cur.r);
      if (qBend || rBend) this.bends.push(cur);
    }
  }

  reducePoints() {
    //reduce tiles (not first or last)
    /*
        let reduced = [];
        console.log(drop);
        for (i = 0; i < this.tiles.length; i++) {
            if (drop.indexOf(i) !== -1)
                continue;
            reduced.push(this.tiles[i]);
        }
        this.tiles = reduced;
         */
  }

  getPoints(layout) {
    let points = 'M';
    for (var i = 0; i < this.tiles.length; i++) {
      let hex = this.tiles[i];
      if (this._isBend(hex)) {
        //cubic curve
        let prev = HexUtils.hexToPixel(this.tiles[i - 1], layout);
        let point = HexUtils.hexToPixel(this.tiles[i], layout);
        let next = HexUtils.hexToPixel(this.tiles[i + 1], layout);

        let newBefore = {};
        newBefore.x = prev.x + (point.x - prev.x) / 2;
        newBefore.y = prev.y + (point.y - prev.y) / 2;
        points += ` ${newBefore.x},${newBefore.y} `;
        //points = points.substr(0, points.length-1);
        points += `Q ${point.x},${point.y} `;

        let newAfter = {};
        newAfter.x = next.x + (point.x - next.x) / 2;
        newAfter.y = next.y + (point.y - next.y) / 2;
        points += ` ${newAfter.x},${newAfter.y} L`;
      } else {
        let p = HexUtils.hexToPixel(hex, layout);
        points += ` ${p.x},${p.y} L`;
      }
    }
    return points.substr(0, points.length - 1);
  }

  _isBend(hex) {
    if (this.bends.length === 0) return false;

    for (var i = 0; i < this.bends.length; i++) {
      if (this.bends[i].q === hex.q && this.bends[i].r === hex.r) return true;
    }

    return false;
  }

  getSpeed() {
    return this.tiles.length - this.bends.length;
  }
}

export default TilePath;
